### ISSUE TITLE
NS26: `Error registering to poll` for guest vote

### ISSUE THREAD

**Issue by adripo** [nextcloud/polls#2971]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### What went wrong, what did you observe?

When sharing any type of poll in NC26 with public link for guest users to vote, there is an error when registering to poll.

The user is unable to register and vote the poll, although the user is created in the sharing details and have its own link. 

![image](https://github.com/nextcloud/polls/assets/26493496/31f6cccf-16b9-4c65-95d1-badd634f7de9)

![image](https://github.com/nextcloud/polls/assets/26493496/bfe49aca-a3ba-4486-af28-30e7ce41a57b)

Even if I pass this link to the user manually there are other errors that continue to appear when voting `Error saving vote` (but the vote appears to be saved on the backend).

![image](https://github.com/nextcloud/polls/assets/26493496/f27a6efc-617a-4a50-a448-7e33dc9e44f1)

### What did you expect, how polls should behave instead?

Being able to register to vote on poll

### What steps does it need to replay this bug?

1. Add new poll -> Date/Text poll
2. Sharing -> Add new public link
3. Open generated link in incognito
4. Enter nickname
5. See notification error `Error registering to poll`

### Installation method

Installed/updated from the appstore (Apps section of your site)

### Installation type

First time installation

### Affected polls version

5.1.0

### Which browser did you use, when experiencing the bug?

- [X] Firefox
- [X] Chrome
- [X] Chromium/Chromium based (i.e. Edge)
- [ ] Safari
- [ ] Other/Don't know

### Add your browser log here

```shell
REQUEST:

curl 'https://REDACTED/apps/polls/s/REDACTED/register?time=1688165026344' \
  -H 'authority: REDACTED' \
  -H 'accept: application/json' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'content-type: application/json' \
  -H 'cookie: REDACTED' \
  -H 'nc-polls-client-id: REDACTED' \
  -H 'origin: REDACTED' \
  -H 'requesttoken: REDACTED' \
  -H 'sec-ch-ua: "Chromium";v="112", "Microsoft Edge";v="112", "Not:A-Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-origin' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36 Edg/112.0.1722.58' \
  --data-raw '{"userName":"test_user2","emailAddress":""}' \
  --compressed


RESPONSE:
500 Internal Server Error
```

### NC version

Nextcloud 26 (Nextcloud Hub 4)

### Other Nextcloud version

26.0.3

### PHP engine version

PHP 8.1

### Other PHP version

8.2

### Database engine

PostgreSQL

### Database Engine version or other Database

15.x

### Which user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other/Don't know

### Add your nextcloud server log here

```shell
Registration logs:

[index] Error: OCP\AppFramework\Db\DoesNotExistException: Did expect one result but found none when executing: query "SELECT * FROM `*PREFIX*polls_watch` WHERE (`poll_id` = :dcValue1) AND (`table` = :dcValue2) AND (`session_id` = :dcValue3)";  at <<closure>>

 0. /var/www/html/lib/public/AppFramework/Db/QBMapper.php line 362
    OCP\AppFramework\Db\QBMapper->findOneQuery(["OC\\DB\\QueryBuilder\\QueryBuilder"])
 1. /var/www/html/custom_apps/polls/lib/Db/WatchMapper.php line 83
    OCP\AppFramework\Db\QBMapper->findEntity(["OC\\DB\\QueryBuilder\\QueryBuilder"])
 2. /var/www/html/custom_apps/polls/lib/Service/WatchService.php line 97
    OCA\Polls\Db\WatchMapper->findForPollIdAndTable(6, "shares")
 3. /var/www/html/custom_apps/polls/lib/Listener/BaseListener.php line 192
    OCA\Polls\Service\WatchService->writeUpdate(6, "shares")
 4. /var/www/html/custom_apps/polls/lib/Listener/BaseListener.php line 101
    OCA\Polls\Listener\BaseListener->writeWatch()
 5. /var/www/html/lib/private/EventDispatcher/ServiceEventListener.php line 86
    OCA\Polls\Listener\BaseListener->handle(["OCA\\Polls\\Ev ... "])
 6. /var/www/html/3rdparty/symfony/event-dispatcher/EventDispatcher.php line 251
    OC\EventDispatcher\ServiceEventListener->__invoke(["OCA\\Polls\\Ev ... "], "OCA\\Polls\\Eve ... t", ["Symfony\\Compo ... "])
 7. /var/www/html/3rdparty/symfony/event-dispatcher/EventDispatcher.php line 73
    Symfony\Component\EventDispatcher\EventDispatcher->callListeners([["Closure"]], "OCA\\Polls\\Eve ... t", ["OCA\\Polls\\Ev ... "])
 8. /var/www/html/lib/private/EventDispatcher/EventDispatcher.php line 87
    Symfony\Component\EventDispatcher\EventDispatcher->dispatch(["OCA\\Polls\\Ev ... "], "OCA\\Polls\\Eve ... t")
 9. /var/www/html/lib/private/EventDispatcher/EventDispatcher.php line 99
    OC\EventDispatcher\EventDispatcher->dispatch("OCA\\Polls\\Eve ... t", ["OCA\\Polls\\Ev ... "])
10. /var/www/html/custom_apps/polls/lib/Service/ShareService.php line 243
    OC\EventDispatcher\EventDispatcher->dispatchTyped(["OCA\\Polls\\Ev ... "])
11. /var/www/html/custom_apps/polls/lib/Controller/PublicController.php line 324
    OCA\Polls\Service\ShareService->register(["OCA\\Polls\\Db\\Share",27], "test_user2", "", "")
12. /var/www/html/custom_apps/polls/lib/Controller/BaseController.php line 87
    OCA\Polls\Controller\PublicController->OCA\Polls\Controller\{closure}("*** sensitive parameters replaced ***")
13. /var/www/html/custom_apps/polls/lib/Controller/PublicController.php line 323
    OCA\Polls\Controller\BaseController->responseCreate(["Closure"], "REDACTED")
14. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 230
    OCA\Polls\Controller\PublicController->register("REDACTED", "test_user2", "", "")
15. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 137
    OC\AppFramework\Http\Dispatcher->executeController(["OCA\\Polls\\Co ... "], "register")
16. /var/www/html/lib/private/AppFramework/App.php line 183
    OC\AppFramework\Http\Dispatcher->dispatch(["OCA\\Polls\\Co ... "], "register")
17. /var/www/html/lib/private/Route/Router.php line 315
    OC\AppFramework\App::main("OCA\\Polls\\Con ... r", "register", ["OC\\AppFramewo ... "], ["REDACTED","polls.public.register"])
18. /var/www/html/lib/base.php line 1060
    OC\Route\Router->match("/apps/polls/s/REDACTED/register")
19. /var/www/html/index.php line 36
    OC::handleRequest()

POST /apps/polls/s/REDACTED/register?time=1688165026344
from REDACTED at REDACTED
```

### Additional environment informations

```shell
Logs related to voting:

[index] Error: OCP\AppFramework\Db\DoesNotExistException: Did expect one result but found none when executing: query "SELECT * FROM `*PREFIX*polls_watch` WHERE (`poll_id` = :dcValue1) AND (`table` = :dcValue2) AND (`session_id` = :dcValue3)";  at <<closure>>

 0. /var/www/html/lib/public/AppFramework/Db/QBMapper.php line 362
    OCP\AppFramework\Db\QBMapper->findOneQuery(["OC\\DB\\QueryBuilder\\QueryBuilder"])
 1. /var/www/html/custom_apps/polls/lib/Db/WatchMapper.php line 83
    OCP\AppFramework\Db\QBMapper->findEntity(["OC\\DB\\QueryBuilder\\QueryBuilder"])
 2. /var/www/html/custom_apps/polls/lib/Service/WatchService.php line 97
    OCA\Polls\Db\WatchMapper->findForPollIdAndTable(6, "votes")
 3. /var/www/html/custom_apps/polls/lib/Listener/BaseListener.php line 192
    OCA\Polls\Service\WatchService->writeUpdate(6, "votes")
 4. /var/www/html/custom_apps/polls/lib/Listener/BaseListener.php line 101
    OCA\Polls\Listener\BaseListener->writeWatch()
 5. /var/www/html/lib/private/EventDispatcher/ServiceEventListener.php line 86
    OCA\Polls\Listener\BaseListener->handle(["OCA\\Polls\\Event\\VoteSetEvent"])
 6. /var/www/html/3rdparty/symfony/event-dispatcher/EventDispatcher.php line 251
    OC\EventDispatcher\ServiceEventListener->__invoke(["OCA\\Polls\\Event\\VoteSetEvent"], "OCA\\Polls\\Event\\VoteSetEvent", ["Symfony\\Compo ... "])
 7. /var/www/html/3rdparty/symfony/event-dispatcher/EventDispatcher.php line 73
    Symfony\Component\EventDispatcher\EventDispatcher->callListeners([["Closure"]], "OCA\\Polls\\Event\\VoteSetEvent", ["OCA\\Polls\\Event\\VoteSetEvent"])
 8. /var/www/html/lib/private/EventDispatcher/EventDispatcher.php line 87
    Symfony\Component\EventDispatcher\EventDispatcher->dispatch(["OCA\\Polls\\Event\\VoteSetEvent"], "OCA\\Polls\\Event\\VoteSetEvent")
 9. /var/www/html/lib/private/EventDispatcher/EventDispatcher.php line 99
    OC\EventDispatcher\EventDispatcher->dispatch("OCA\\Polls\\Event\\VoteSetEvent", ["OCA\\Polls\\Event\\VoteSetEvent"])
10. /var/www/html/custom_apps/polls/lib/Service/VoteService.php line 137
    OC\EventDispatcher\EventDispatcher->dispatchTyped(["OCA\\Polls\\Event\\VoteSetEvent"])
11. /var/www/html/custom_apps/polls/lib/Controller/PublicController.php line 187
    OCA\Polls\Service\VoteService->set(38, "yes", ["OCA\\Polls\\Model\\Acl"])
12. /var/www/html/custom_apps/polls/lib/Controller/BaseController.php line 55
    OCA\Polls\Controller\PublicController->OCA\Polls\Controller\{closure}("*** sensitive parameters replaced ***")
13. /var/www/html/custom_apps/polls/lib/Controller/PublicController.php line 186
    OCA\Polls\Controller\BaseController->response(["Closure"], "REDACTED")
14. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 230
    OCA\Polls\Controller\PublicController->setVote(38, "yes", "REDACTED")
15. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 137
    OC\AppFramework\Http\Dispatcher->executeController(["OCA\\Polls\\Co ... "], "setVote")
16. /var/www/html/lib/private/AppFramework/App.php line 183
    OC\AppFramework\Http\Dispatcher->dispatch(["OCA\\Polls\\Co ... "], "setVote")
17. /var/www/html/lib/private/Route/Router.php line 315
    OC\AppFramework\App::main("OCA\\Polls\\Con ... r", "setVote", ["OC\\AppFramewo ... "], ["REDACTED","polls.public.set_vote"])
18. /var/www/html/lib/base.php line 1060
    OC\Route\Router->match("/apps/polls/s/REDACTED/vote")
19. /var/www/html/index.php line 36
    OC::handleRequest()

PUT /apps/polls/s/REDACTED/vote
from REDACTED at REDACTED

```

### Additional Information

`Hide login option in public polls` is enabled in Poll Admin settings.

Already tried `occ polls:db:rebuild` but didn't solve.
Note: all my old polls are broken. There is only one option visible and no votes registered.

**Comment by dartcafe** [comment]
I can't reproduce it, but it seems, there is something wrong inside the tables. 

What happens: 
1. Polls wants to update the information about the last updated table into `oc_polls_watch`.
2. `REASON_UNIQUE_CONSTRAINT_VIOLATION` is thrown then (Entry exists) and gets catched
   This means: The entry for this table already exists
4. Then Polls wants to read the existing entry to update it, but then it fails with `DoesNotExistException`
   This means: the entry for this table does NOT exist

I am confused! Thsi makes no sense.

Are you able to send me a DB dump (the current and maybe a backup from before the update to 5.1?

https://nc.yagst.de/s/CFfosAW68CjZt9H

**Comment by adripo** [comment]
Sorry, but I am not able to dump the DB for privacy concerns. Is there any command that I can execute to remove all the tables and recreate them empty? Could this solve the issue? It is strange that even the newly creates polls continues to have issues.

**Comment by dartcafe** [comment]
Without data, I can't analyse the issue.

**Comment by adripo** [comment]
I used `polls:db:purge` and restored eveything clean. Now the app works correctly. Something must have happened during the upgrade. I am closing the issue.

**Comment by fsamapoor** [comment]
@dartcafe @adripo 

Had the exact same problem with 5.2.0.

In fresh installs, everything is alright, but somehow the indices on the `oc_polls_watch` have not been upgraded with older instances.

We have an upgraded instance from v24 to v26.

These are the indices on the fresh install for the `oc_polls_watch` table:

![Screenshot from 2023-11-23 09-41-59](https://github.com/nextcloud/polls/assets/4992968/35b29270-4873-4d36-b059-ead35b22eb7f)

These are the indices on the upgraded instance for the `oc_polls_watch` table:

![Screenshot from 2023-11-23 09-41-34](https://github.com/nextcloud/polls/assets/4992968/194ee578-fc24-449b-88b0-f4c875fe5c20)

Basically, the `uniq_watch` index is causing the problem because with upgraded instances that have been using the app, there is always an entry matching the `table` and `column_id` but when it tries to update it, the `session_id` is not the same as the existing entry, causing the logic to throw an exception.

I'm not sure what's the best way to ensure that this index gets updated with the upgrades, as the approach to dealing with migrations does not look familiar to me.

**Comment by dartcafe** [comment]
please first update your polls version to the current release. If this problem still exists, open a new bug report and fill out the required fields.
